src/mesh/mesh_adaptation/anisotropic_mesh_adaptation.cpp

```diff
@@ -0,0 +1,537 @@
+    PHiLiP::FEValuesShapeHessian<dim> fe_values_shape_hessian;
+    
+    std::vector<dealii::types::global_dof_index> dof_indices(max_dofs_per_cell);
+
+    for(const auto &cell : dg->dof_handler.active_cell_iterators())
+    {
+        if(! cell->is_locally_owned()) {continue;}
+        
+        const unsigned int cell_index = cell->active_cell_index();
+        const unsigned int i_fele = cell->active_fe_index();
+        const unsigned int i_quad = i_fele;
+        const unsigned int i_mapp = 0;
+        fe_values_collection_volume.reinit(cell, i_quad, i_mapp, i_fele);
+        const dealii::FEValues<dim,dim> &fe_values_volume = fe_values_collection_volume.get_present_fe_values();
+        
+        const unsigned int n_dofs_cell = fe_values_volume.dofs_per_cell; 
+        dof_indices.resize(n_dofs_cell);
+        cell->get_dof_indices(dof_indices);
+    
+        cellwise_hessian[cell_index] = 0;
+        const unsigned int iquad = get_iquad_near_cellcenter(fe_values_volume.get_quadrature());
+        fe_values_shape_hessian.reinit(fe_values_volume, iquad);
+        const dealii::FESystem<dim,dim> &fe_ref = fe_values_volume.get_fe();
+        
+        for(unsigned int idof = 0; idof<n_dofs_cell; ++idof)
+        {
+            const unsigned int icomp = fe_values_volume.get_fe().system_to_component_index(idof).first;
+            // Computing hesssian of solution at state 0. Might need to change it later.
+            if(icomp == 0)
+            {
+                cellwise_hessian[cell_index] += dg->solution(dof_indices[idof])*fe_values_shape_hessian.shape_hessian_component(idof, iquad, icomp, fe_ref);
+            }
+        }
+        cellwise_hessian[cell_index] = get_positive_definite_tensor(cellwise_hessian[cell_index]);
+    }
+    
+    change_p_degree_and_interpolate_solution(initial_poly_degree);
+    dg->solution = solution_old; // reset solution
+    dg->solution.update_ghost_values();
+}
+
+template<int dim, int nstate, typename real, typename MeshType>
+void AnisotropicMeshAdaptation<dim, nstate, real, MeshType> :: compute_goal_oriented_hessian()
+{
+    // Compute goal oriented pseudo hessian.
+    // From Eq. 28 in Loseille, A., Dervieux, A., and Alauzet, F. "Fully anisotropic goal-oriented mesh adaptation for 3D steady Euler equations.", 2010.
+    // Also, metric terms related to faces do not make a difference and hence are not included (ref: footnote on page 78 in  
+    // Dervieux, A., Alauzet, F., Loseille, A., Koobus, B. (2022). Mesh Adaptation for Computational Fluid Dynamics 2.
+    
+    // Compute the adjoint ===================================================================
+    VectorType adjoint(dg->solution); 
+    dg->assemble_residual(true);
+    functional->evaluate_functional(true);
+    solve_linear(dg->system_matrix_transpose, functional->dIdw, adjoint, dg->all_parameters->linear_solver_param);
+    adjoint *= -1.0;
```

Comment:
Should the adjoint be computed on the fine space?

Reply:
Yes, mathematically it should be on the fine space (especially for DG schemes due to Galerkin Orthogonality of the PDE to the coarse polynomial space). Fixed in the recent commit.